README
GMapCatcher - An offline map viewer

Clone of the original project, which can be found here:
GMapCatcher - An offline map viewer
HomePage: http://code.google.com/p/gmapcatcher/



changelog
January 28 2015    M.D    <draeger.marco@gmail.com>
    * 0.8.1.0
    * Added ESRI ArcGIS tile cache export (exploded)

July 20 2013    H.S    <HelderSepu@gmail.com>
    * 0.8.0.5 - At revision: 1590
    * Fix issue in py2exe that created problems in Win 8

July 19 2013    K.H    <kipenroskaposti@gmail.com>
    * 0.8.0.4 - At revision: 1585
    * Update gpxpy

July 18 2013    H.S    <HelderSepu@gmail.com>
    * 0.8.0.3 - At revision: 1581
    * Fix issue with Stamen Maps
    * Fix issue with google searches
    * Start support for Win 8

February 09 2013    H.S    <HelderSepu@gmail.com>
    * 0.8.0.2 - At revision: 1568
    * Draw markers in a real thread. (Issue 347)
    * Fix issue with Nokia Maps

October 12 2012    H.S    <HelderSepu@gmail.com>
    * 0.8.0.0 - At revision: 1524
    * Read GPS NMEA sentences (Issue 77)
    * Export GPS traces (GPX) (Issue 79)
    * Avoid restarting GMapCatcher when GPSd restarts (Issue 98)
    * Draw tracks on the map (Issue 101)
    * Get Directions (Issue 103)
    * Add Check box to download Terrain,Satellite or Map (Issue 109)
    * Create one db file per map service. (Issue 207)
    * Add new GPSd protocol (Issue 209) 
    * Fix Incorrect scale (Issue 281)
    * Fix GPS heading arrow (Issue 332)
    * Fix issue with markers when given same name (Issue 335)
    * Add Nokia maps (Issue 336)
    * Add skyvector maps (Issue 337)
    * Show more GPS data (Issue 338)

April 28 2012    H.S    <HelderSepu@gmail.com>
    * 0.7.7.2 - At revision: 1240
    * Add measuring method to find the distance between two points (Issue 277)
    * Add config option to set API for CloudMade maps (Issue 306)
    * Fix error using draw_line (Issue 321)
    * Add support for Stamen maps: http://maps.stamen.com
    
March 15 2012    H.S    <HelderSepu@gmail.com>
    * 0.7.7.1 - At revision: 1231
    * Add the ability to change the font/style/color of the marker

Feb 27 2012    H.S    <HelderSepu@gmail.com>
    * 0.7.7.0 - At revision: 1225
    * OpenCycleMap tiles server changed (Issue 308)
    * Added OSM format (Issue 300)
    * Download tiles from a GPX path (Issue 289)
    * Import/Export maps (Issue 92)
    * Change the ".googlemaps" folder to ".GMapCatcher" (Issue 276)

May 1 2011    H.S    <HelderSepu@gmail.com>
    * 0.7.6.1 - At revision: 1138
    * Fix issue with the search (Issue 264)

March 28 2011    H.S    <HelderSepu@gmail.com>
    * 0.7.6.0 - At revision: 1130
    * Upgrade windows to use Python 2.7
    * Remove Google from the list of servers, set default to Yahoo (Issue 210)
    * Add support for yandex maps (Issue 34)
    * Show GPS direction (Issue 78)
    * Move map to entered coordinates (Issue 191)
    * Add support for MGMaps format file storage (Issue 202)
    * Add GPS update timeout (Issue 213)
    * Improved "Force update" function (Issue 216)
    * Add support for mapy.cz (Issue 220)
    * Improve Unicode support (Issue 223)
    * Add command line parameter (--config-path="gmapcatcher.conf") (Issue 227)
    * Improve the auto-completion for the entry box (Issue 238)

August 22 2010    H.S    <HelderSepu@gmail.com>
    * 0.7.5.0 - At revision: 986
    * Complete the Export Image tool (Issue 99)
    * Fix Google's URL pattern (Issues 197, 198, 199)
    * Add Yahoo Hybrid
    * Change Download Queue to "Last In First Out" (Issue 204)
    * Fix zoom level after search (Issue 206)

July 21 2010    M.B    <MarkieB.lists.20090330@gmail.com>
    * 0.7.2.2 - At revision: 928
    * Save window size/location/layer at close
    * Add gtk window ico
    * Debian build
    * Mark up map center without image file needed
    * Change onedirpermap / map service provider without restart
    * Add possibility for distributions to demand strict legal adherence

July 14 2010    H.S    <HelderSepu@gmail.com>
    * 0.7.2.0 - At revision: 881
    * Display coordinates in StatusBar (Issue 41)
    * Feedback to the user when maps are downloading (Issue 142)
    * Allow google Hybrid maps (Issue 174)
    * Manual tile download from GUI (Issue 182)

July 06 2010    H.S    <HelderSepu@gmail.com>
    * 0.7.1.0 - At revision: 839
    * Change the Google terrain pattern (Issue 175)

April 17 2010    H.S    <HelderSepu@gmail.com>
    * 0.7.0.0 - At revision: 816
    * Add option to separate file paths for different map sources. (Issue 120)
    * Fix bug downloading big sections. (Issue 135)
    * Catch the map of China without english name. (Issue 136)
    * Add message for Mass download from OpenCycleMap. (Issue 145)
    * Fix bug searching locations within China. (Issue 150)

March 4 2010    N.W    <pi3orama@gmail.com>
    * Rearrange the 'Download' button and map server combobox
    * Add a language setting for google map

February 15 2010    H.S    <HelderSepu@gmail.com>
    * 0.6.3.0 - At revision: 741
    * Fix bug loading KML files (Issue 124)

January 30 2010    H.S    <HelderSepu@gmail.com>
    * 0.6.2.0 - At revision: 717
    * Add (Key + Click) functionality (Issue 123)
      - Ctrl + Click = Add Marker
      - Alt + 2Click = Zoom Out

January 8 2010    H.S    <HelderSepu@gmail.com>
    * 0.6.1.0 - At revision: 670
    * Fix bug Importing markers from KML files (Issue 65)

January 5 2010    H.S    <HelderSepu@gmail.com>
    * 0.6.0.0 - At revision: 662
    * Use Google's new URL pattern (Issue 112)
    * Add support for VirtualEarth (Issue 114)
    * Import markers from KML files (Issue 65)
    * Add a "Add Marker" to the right Click menu (Issue 65)

November 21 2009    H.S    <HelderSepu@gmail.com>
    * 0.5.0.0 - At revision: 622
    * Fix download of big areas (Issue 100)
    * Keep selected location visible (Issue 102)
    * Improve look in Ubuntu 9.10 (Issue 108)

October 30 2009    H.S    <HelderSepu@gmail.com>
    * 0.4.0.0 - At revision: 592
    * Add more map services (Issue 82)
    * Fix bug in the Path to images (Issue 95)
    * Apply changes to the markers immediately (Issue 97)

October 11 2009    H.S    <HelderSepu@gmail.com>

gmapcatcher/mapTilesTransfer.py
# -*- coding: utf-8 -*-
## @package gmapcatcher.mapTilesTransfer
# Transfer tiles from repos 1 to repos 2
# EDIT: 2015 Marco Draeger
#       Added distinction for ArcGISExploded Repo to provide information for layer

import threading
import time
import gobject

import mapUtils
from tilesRepo.tilesRepo import TilesRepository
from tilesRepo.tilesRepoArcGISExplode import TilesRepositoryArcGISExplode


class InvalidInputParametersError(Exception):
    pass


class TilesTransfer(threading.Thread):

    def __init__(self, trepos_source, trepos_destination, center, zooms, region, layer, overwrite_destination):
        """copy tiles from repos trepos_source to trepos_destination

        trepos_source: source repository
        trepos_destination: destination repository
        center: (lat, lon)
        zooms: (zoom_min, zoom_max)
        region: (width, height) [km]
        layer: what layer to transfer
        """

        threading.Thread.__init__(self)
        self.stop_lock = threading.Lock()

        # when set to True, thread should stop immediately
        self.stop = False

        self.trepos_source = trepos_source
        self.trepos_destination = trepos_destination

        self.center_lat = center[0]
        self.center_lon = center[1]
        self.zoom_min = int(zooms[0])
        self.zoom_max = int(zooms[1])
        self.reg_width = region[0]
        self.reg_height = region[1]
        self.layer = layer
        self.overwrite_destination = overwrite_destination

        if (not isinstance(trepos_source, TilesRepository)):
            raise InvalidInputParametersError("trepos_source is not subclass of TilesRepository")
        if (not isinstance(trepos_destination, TilesRepository)):
            raise InvalidInputParametersError("trepos_destination is not subclass of TilesRepository")
        if self.zoom_max < self.zoom_min:
            raise InvalidInputParametersError("Zoom max (%d) is less than zoom min (%d)." % (self.zoom_max, self.zoom_min))

    def dlon_dlat(self):
        # calculate the height and widt of the exported region
        dlon = mapUtils.km_to_lon(mapUtils.nice_round(self.reg_width), self.center_lat)
        dlat = mapUtils.km_to_lat(mapUtils.nice_round(self.reg_height))

        if dlat > 170:
            dlat = 170
        if dlon > 358:
            dlon = 358

        return (dlon, dlat)

    def get_tiles_range_for_zoom(self, zoom):
        # get tiles - copied from mapDownloader

        (dlon, dlat) = self.dlon_dlat()

        top_left = mapUtils.coord_to_tile(
            (self.center_lat + dlat / 2, self.center_lon - dlon / 2, zoom)
        )
        bottom_right = mapUtils.coord_to_tile(
            (self.center_lat - dlat / 2, self.center_lon + dlon / 2, zoom)
        )

        # top_left[0][0], bottom_right[0][0], top_left[0][1], bottom_right[0][1]
        # xmin, xmax, ymin, ymax

        world_tiles = mapUtils.tiles_on_level(zoom)
        if bottom_right[0][0] - top_left[0][0] >= world_tiles:
            top_left[0][0], bottom_right[0][0] = 0, world_tiles - 1
        if bottom_right[0][1] - top_left[0][1] >= world_tiles:
            top_left[0][1], bottom_right[0][1] = 0, world_tiles - 1

        # xmin, xmax, ymin, ymax
        return(top_left[0][0], bottom_right[0][0], top_left[0][1], bottom_right[0][1])

    def set_callback_update(self, callback):
        self.callback_update = callback

    def set_callback_finish(self, callback):
        self.callback_finish = callback

    def count_all_tiles(self):
        all_tiles = 0

        zoom = self.zoom_min
        while zoom <= self.zoom_max:
            tiles_range = self.get_tiles_range_for_zoom(zoom)
            all_tiles = all_tiles + (tiles_range[1] - tiles_range[0] + 1) * (tiles_range[3] - tiles_range[2] + 1)
            zoom = zoom + 1

        return all_tiles

    def run(self):
        """Do transfer tiles.

        if overwrite is true, overwrite existing tiles in destination repository
        """

        if (isinstance (self.trepos_destination, TilesRepositoryArcGISExplode)):
            (dlon, dlat) = self.dlon_dlat()
            xmin = self.center_lon - dlon/2
            ymin = self.center_lat - dlat/2
            xmax = self.center_lon + dlon/2
            ymax = self.center_lat + dlat/2

            self.trepos_destination.writeConfig(xmin, ymin, xmax, ymax, self.layer)

        gobject.idle_add(self.callback_update, "Computing tiles...")
        num_all_tiles = self.count_all_tiles()

        update_time = time.time()

        zoom = self.zoom_max

        tiles_count = 0
        tiles_written_count = 0
        while zoom >= self.zoom_min:
            tiles_range = self.get_tiles_range_for_zoom(zoom)

            ty = tiles_range[2]
            while ty <= tiles_range[3]:
                tx = tiles_range[0]
                while tx <= tiles_range[1]:
                    if self.should_i_stop():
                        return

                    if self.trepos_source.is_tile_in_local_repos((tx, ty, zoom), self.layer):
                        do_it = True

                        if not self.overwrite_destination:
                            if self.trepos_destination.is_tile_in_local_repos((tx, ty, zoom), self.layer):
                                do_it = False

                        if do_it:
                            tile_data = self.trepos_source.get_plain_tile((tx, ty, zoom), self.layer)
                            self.trepos_destination.store_plain_tile((tx, ty, zoom), self.layer, tile_data)
                            tiles_written_count = tiles_written_count + 1

                    tx = tx + 1
                    tiles_count = tiles_count + 1

                    if time.time() - update_time > 1:
                        percent = int((tiles_count / (1.0 * num_all_tiles)) * 100)
                        text = "Processed %d%% (%d of %d tiles)" % (percent, tiles_count, num_all_tiles)
                        gobject.idle_add(self.callback_update, text, percent)
                        update_time = time.time()

                ty = ty + 1

            zoom = zoom - 1
        gobject.idle_add(self.callback_finish, "All %d tiles processed." % (num_all_tiles, ))

    def should_i_stop(self):
        self.stop_lock.acquire()
        ret = self.stop
        self.stop_lock.release()
        return ret

    def set_stop(self, value):
        self.stop_lock.acquire()
        self.stop = value
        self.stop_lock.release()


gmapcatcher/tilesRepo/tilesRepoArcGISExplode.py
# -*- coding: utf-8 -*-
## @package gmapcatcher.tilesRepo.tilesRepoArcGISExplode
# This modul provides filebased tile repository functions
#
# Usage:
#
# - constructor requires MapServ instance, because method
#  'get_tile_from_coord' is provided in the MapServ
#
# tilesRepoArcGISExplode exports tiles in ArcGIS cache format
# Copyright (C) 2015  Marco Draeger
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import os
import gmapcatcher.lrucache as lrucache
import gmapcatcher.fileUtils as fileUtils
import gmapcatcher.widgets.mapPixbuf as mapPixbuf
import xml.dom.minidom

from threading import Lock
from gmapcatcher.mapConst import *
from tilesRepo import TilesRepository
from tilesRepoSQLite3 import tileNotInRepository


class TilesRepositoryArcGISExplode(TilesRepository):

    def __init__(self, MapServ_inst, conf):
        TilesRepository.__init__(self, MapServ_inst, conf)
        self.configpath = conf.init_path
        self.tile_cache = lrucache.LRUCache(1000)
        self.mapServ_inst = MapServ_inst
        self.lock = Lock()
        self.missingPixbuf = mapPixbuf.missing()

    def writeConfig(self, xmin, ymin, xmax, ymax, layer):
        path = os.path.join(self.configpath, "ArcGIS_" + LAYER_DIRS[layer])
        path = fileUtils.check_dir(path)
        self.writeConfCDI(os.path.join(path, 'conf.cdi'), xmin, ymin, xmax, ymax)
        self.writeConfXML(os.path.join(path, 'conf.xml'))

    def writeConfCDI(self, filename, xmin, ymin, xmax, ymax):
        domString = """<EnvelopeN xsi:type='typens:EnvelopeN' xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance' xmlns:xs='http://www.w3.org/2001/XMLSchema' xmlns:typens='http://www.esri.com/schemas/ArcGIS/10.1'><XMin>%.6f</XMin><YMin>%.6f</YMin><XMax>%.6f</XMax><YMax>%.6f</YMax></EnvelopeN>""" % (xmin, ymin, xmax, ymax)
        dom = xml.dom.minidom.parseString(domString)
        f = open(filename, 'wb')
        dom.writexml(f)
        f.close()

    def writeConfXML(self, filename):
        domString = """<CacheInfo xmlns:typens="http://www.esri.com/schemas/ArcGIS/10.1" xmlns:xs="http://www.w3.org/2001/XMLSchema" 
                                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="typens:CacheInfo">
                          <TileImageInfo xsi:type="typens:TileImageInfo">
                             <CacheTileFormat>PNG</CacheTileFormat>
                          </TileImageInfo>
                       </CacheInfo>"""
        dom = xml.dom.minidom.parseString(domString)
        f = open(filename, 'wb')
        dom.writexml(f)
        f.close()

    def finish(self):
        # last command in finish
        TilesRepository.finish(self)

    ## Sets new repository path to be used for storing tiles
    def set_repository_path(self, conf):
        self.configpath = conf.init_path

    ## check if we have locally downloaded tile
    def is_tile_in_local_repos(self, coord, layer):
        path = self.coord_to_path(coord, layer)
        return  os.path.isfile(path)

    ## Returns the PixBuf of the tile
    #  Uses a cache to optimise HDD read access
    def load_pixbuf(self, coord, layer, force_update):
        filename = self.coord_to_path(coord, layer)
        if not force_update and (filename in self.tile_cache):
            pixbuf = self.tile_cache[filename]
        else:
            if os.path.isfile(filename):
                try:
                    pixbuf = mapPixbuf.image_data_fs(filename)
                    self.tile_cache[filename] = pixbuf
                except Exception:
                    pixbuf = self.missingPixbuf
                    print "File corrupted: %s" % filename
                    fileUtils.del_file(filename)
            else:
                pixbuf = self.missingPixbuf
        return pixbuf

    ## Get the png file for the given location
    #  Returns true if the file is successfully retrieved
    def get_png_file(self, coord, layer,
                        online, force_update, conf):
        filename = self.coord_to_path(coord, layer)
        # remove tile only when online
        remove_tile = False
        if (force_update and online):
            remove_tile = fileUtils.is_old(filename, conf.force_update_days)

        if os.path.isfile(filename) and not remove_tile:
            return True
        if not online:
            return False

        try:
            data = self.mapServ_inst.get_tile_from_coord(
                        coord, layer, conf
                    )
            self.coord_to_path_checkdirs(coord, layer)
            # Remove the old tile only after getting the new data
            if remove_tile:
                fileUtils.delete_old(filename, conf.force_update_days)
            file = open(filename, 'wb')
            file.write(data)
            file.close()

            return True
        except KeyboardInterrupt:
            raise
        except Exception, excInst:
            print excInst
        return False

    def get_plain_tile(self, coord, layer):
        if not self.is_tile_in_local_repos(coord, layer):
            raise tileNotInRepository(str((coord, layer)))

        filename = self.coord_to_path(coord, layer)
        thefile = open(filename, 'rb')
        ret = thefile.read()
        thefile.close()
        return ret

    def store_plain_tile(self, coord, layer, tiledata):
        filename = self.coord_to_path_checkdirs(coord, layer)
        file = open(filename, 'wb')
        file.write(tiledata)
        file.close()

    ## Return the absolute path to a tile
    #  only check path
    #  tile_coord = (zoom_level, tile_Y, tile_X)
    #  sample of the Naming convention:
    #  \.googlemaps\ArcGIS\_allLayers\L00\R0000002\C0000001.png
    #  We only have 2 levels for one axis
    # private
    def coord_to_path(self, tile_coord, layer):
        column = 'C' + ("%08x" % tile_coord[0]).lower()
        row    = 'R' + ("%08x" % tile_coord[1]).lower()
        zoom   = 'L' + ("%02d" % (MAP_MAX_ZOOM_LEVEL - tile_coord[2]))
        return os.path.join(
                            self.configpath,
                            "ArcGIS_" + LAYER_DIRS[layer],
                            "_alllayers",
                            zoom, row, column + ".png"
                           )

    ## create path if doesn't exists
    #  tile_coord = (zoom_level, tile_Y, tile_X)
    #  sample of the Naming convention:
    #  \.googlemaps\ArcGIS\_allLayers\L00\R0000002\C0000001.png
    #  We only have 2 levels for one axis
    # private
    def coord_to_path_checkdirs(self, tile_coord, layer):
        column = 'C' + ("%08x" % tile_coord[0]).lower()
        row    = 'R' + ("%08x" % tile_coord[1]).lower()
        zoom   = 'L' + ("%02d" % (MAP_MAX_ZOOM_LEVEL - tile_coord[2]))
        self.lock.acquire()
        path = os.path.join(self.configpath, "ArcGIS_" + LAYER_DIRS[layer])
        path = fileUtils.check_dir(path)
        path = fileUtils.check_dir(path, "_alllayers")
        path = fileUtils.check_dir(path, zoom)
        path = fileUtils.check_dir(path, row)
        self.lock.release()
        return os.path.join(path, column + '.png')


gmapcatcher/mapConst.py
# -*- coding: utf-8 -*-
## @package gmapcatcher.mapConst
# Place to keep all constants
# EDIT: 2015 Marco Draeger
#       added constant for ArcGIS cache export

import os
import locale

NAME = "GMapCatcher"
VERSION = "0.8.1.0"
VERSION_NAME = ""
WEB_ADDRESS = "http://github.com/mdraeger/gmapcatcher"

GOOGLE = 0
OSM = 1
CLOUDMADE = 2
YAHOO = 3
INFO_FREEWAY = 4
OPENCYCLEMAP = 5
GOOGLE_MAKER = 6
VIRTUAL_EARTH = 7
YANDEX = 8
SEZNAM = 9
SEZNAM_HIKING = 10
SEZNAM_CYCLO = 11
SEZNAM_HIST = 12
STAMEN = 13
REFUGES = 14
OPENSEAMAP = 15
ENIRO = 16
NOKIA = 17
SKYVECTOR_WORLD_VFR = 18
SKYVECTOR_WORLD_LO = 19
SKYVECTOR_WORLD_HI = 20
MAPS_FOR_FREE = 21

MAP_SERVERS = [
    "Google", "OpenStreetMap", "CloudMade", "Yahoo",
    "InformationFreeway", "OpenCycleMap", "Google Map Maker",
    "Virtual Earth", "Yandex",
    "Seznam", "Seznam Turistická", "Seznam Cyklo", "Seznam Historická",
    "Stamen", "Refuges Europe", "OpenSeaMap", "Eniro", "Nokia",
    "SkyVector World VFR", "SkyVector World Lo", "SkyVector World Hi",
    "Maps-For-Free"
]

LAYER_MAP = 0
LAYER_SAT = 1
LAYER_TER = 2
LAYER_HYB = 3
LAYER_CHA = 4
LAYER_NAMES = ["Map", "Satellite", "Terrain", "Hybrid", "Chart"]
LAYER_DIRS = ["tiles", "sat_tiles", "ter_tiles", "hyb_tiles", "chart_tiles"]

FIRST_LAYER_ID = 0
MAP_SERVICES = [
    {"ID": GOOGLE,        "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER, LAYER_HYB]},
    {"ID": OSM,           "layers": [LAYER_MAP]},
    {"ID": CLOUDMADE,     "layers": [LAYER_MAP]},
    {"ID": YAHOO,         "layers": [LAYER_MAP, LAYER_SAT, LAYER_HYB]},
    {"ID": INFO_FREEWAY,  "layers": [LAYER_MAP]},
    {"ID": OPENCYCLEMAP,  "layers": [LAYER_MAP]},
    {"ID": GOOGLE_MAKER,  "layers": [LAYER_MAP]},
    {"ID": VIRTUAL_EARTH, "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER]},
    {"ID": YANDEX,        "layers": [LAYER_MAP]},
    {"ID": SEZNAM,        "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER, LAYER_HYB]},
    {"ID": SEZNAM_HIKING, "layers": [LAYER_SAT, LAYER_TER, LAYER_HYB]},
    {"ID": SEZNAM_CYCLO,  "layers": [LAYER_SAT, LAYER_TER, LAYER_HYB]},
    {"ID": SEZNAM_HIST,   "layers": [LAYER_SAT, LAYER_TER, LAYER_HYB]},
    {"ID": STAMEN,        "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER]},
    {"ID": REFUGES,       "layers": [LAYER_MAP, LAYER_SAT, LAYER_HYB]},
    {"ID": OPENSEAMAP,    "layers": [LAYER_CHA]},
    {"ID": ENIRO,         "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER, LAYER_HYB, LAYER_CHA]},
    {"ID": NOKIA,         "layers": [LAYER_MAP, LAYER_SAT, LAYER_TER]},
    {"ID": SKYVECTOR_WORLD_VFR,  "layers": [LAYER_MAP]},
    {"ID": SKYVECTOR_WORLD_LO,   "layers": [LAYER_MAP]},
    {"ID": SKYVECTOR_WORLD_HI,   "layers": [LAYER_MAP]},
    {"ID": MAPS_FOR_FREE, "layers": [LAYER_TER, LAYER_HYB]},
]

NO_BULK_DOWN = ["Google", "OpenStreetMap", "OpenCycleMap"]
NO_GPS = ["Yahoo"]

MAP_MAX_ZOOM_LEVEL = 17
MAP_MIN_ZOOM_LEVEL = -2
TILES_WIDTH = 256
TILES_HEIGHT = 256
NR_MTS = 4
SEPARATOR = "\t"
TOOLS_MENU = ["Settings", "Edit locations", "Edit markers",
              "Change Theme", "GPS Options", ""]
TOOLS_MENU_PLUS_CREDITS = " About "
TOOLS_MENU_PLUS_VISUAL_DL = "Save Path Maps"

ZOOM_IN = 0
ZOOM_OUT = 1
CENTER_MAP = 2
RESET = 3
BATCH_DOWN = 5
EXPORT_MAP = 6
ADD_MARKER = 7
MOUSE_LOCATION = 9
GPS_LOCATION = 10
GPS_DIRECTIONS = 12
DA_MENU = ["Zoom In", "Zoom Out", "Center map here", "Reset",
        "", "Batch Download", "Export Map", "Add Marker",
        "", "Copy Location", "Copy GPS", "", "Directions from GPS to here"]
DA_MENU_LIM = ["Zoom In", "Zoom Out", "Center map here", "Reset",
        "", "-", "-", "Add Marker", "-", "-", "-", "-", "-"]

REPOS_TYPE_FILES = 0
REPOS_TYPE_SQLITE3 = 1
REPOS_TYPE_MGMAPS = 2
REPOS_TYPE_OSM = 3
REPOS_TYPE_ARCGISEXPLODE = 4
REPOS_TYPE_RMAPS = 5
REPOS_TYPE = ["Files", "SQLite3", "MGMaps", "OSM", "ArcGIS explode", "RMaps"]
DEFAULT_REPOS_TYPE = 0
SQLITE3_REPOSITORY_FILE = "tilerepository.db"
RMAPS_REPOSITORY_FILE_FORMAT = "tile-%s.sqlitedb"

STATUS_NONE = 0
STATUS_GPS = 1
STATUS_MOUSE = 2
STATUS_TYPE = ["None", "GPS", "Mouse"]

GPS_IMG_SIZE = (48, 48)

GPS_DISABLED = 0
GPS_MARKER = 1
GPS_CENTER = 2
GPS_ON_SCREEN = 3
GPS_TIMEOUT = 4
GPS_NAMES = ["GPS Disabled", "GPS Marker", "GPS Center",
             "GPS on Screen", "GPS Timeout"]
TYPE_OFF = 0
TYPE_GPSD = 1
TYPE_SERIAL = 2
GPS_TYPES = ["Off", "GPSd", "Serial"]
MODE_NO_FIX = 1
MODE_2D = 2
MODE_3D = 3
# default distance - fraction of deg - for gps track increment
GPS_INCREMENT = 0.001

# Units for distance measurement and GPS speed
DISTANCE_UNITS = ['km', 'miles', 'NM']
SPEED_UNITS = ['km/h', 'mph', 'knots']
UNIT_TYPE_KM = 0
UNIT_TYPE_MILE = 1
UNIT_TYPE_NM = 2

SECTION_INIT = 'init'
SECTION_GPS = 'gps'
SECTION_MAP = 'map'
SECTION_AGENT = 'agent'
R_EARTH = 6371.

USER_PATH = unicode(os.path.expanduser("~"), locale.getpreferredencoding())
DEFAULT_PATH = os.path.join(USER_PATH, ".GMapCatcher")

LANGUAGES = ["en", "es", "zh"]

STARTS_WITH = 0
ENDS_WITH = 1
CONTAINS = 2
REGULAR_EXPRESSION = 3
ENTRY_SUB_MENU = ['Starts With...', 'Ends With...',
        'Contains...', 'Regular Expression...']

STRICT_LEGAL = False
